platform/darwin/src/MGLFeature.mm

```diff
@@ -188,10 +195,47 @@ - (id)attributeForKey:(NSString *)key {
 }
 
 - (NS_DICTIONARY_OF(NSString *, id) *)featureDictionary {
-    // TODO: MGLShapeCollectionFeature
-    return nil;
+    
+    return @{@"type":@"Feature",
+             @"properties":(self.attributes) ? self.attributes : [NSDictionary dictionary],
+             @"geometry":@{
+                     @"type":@"GeometryCollection",
+                     @"geometries":[self geometryCollection:self.shapes]
+                     }
+             };
 }
 
+- (NS_MUTABLE_ARRAY_OF(NS_DICTIONARY_OF(NSString *, id) *) *)geometryCollection:(NS_ARRAY_OF(MGLShape *) *)shapes {
+    NS_MUTABLE_ARRAY_OF(NS_DICTIONARY_OF(NSString *, id) *) *geometries = [NSMutableArray array];
+    
+    for (MGLShape<MGLFeature> *shape in shapes) {
+        if ([shape isKindOfClass:[MGLPointFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLPointFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLPolylineFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLPolylineFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLPolygonFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLPolygonFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLMultiPointFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLMultiPointFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLMultiPolylineFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLMultiPolylineFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLMultiPolygonFeature class]]) {
+            [self addTypeAndCoordinatesFrom:(MGLMultiPolygonFeature *)shape to:&geometries];
+        } else if ([shape isKindOfClass:[MGLShapeCollectionFeature class]]) {
+            [geometries addObject:@{
+                @"type":@"GeometryCollection",
+                @"geometries":[self geometryCollection:((MGLShapeCollectionFeature *)shape).shapes]}];
+        }
+    }
+    
+    return geometries;
+}
+
+- (void)addTypeAndCoordinatesFrom:(id<MGLFeature>)feature to:(NS_MUTABLE_ARRAY_OF(NS_DICTIONARY_OF(NSString *, id) *)**)geometries {
```

Comment:
This is a questionable refactoring. First of all, since the array is mutable, you can simply pass a pointer to it rather than a double-pointer. Secondly, if all the classes that implement MGLFeature have a `-featureDictionary` method, it would be cleaner to declare that method on MGLFeature. That way, `-geometryCollection:` can simply call `feature.featureDictionary` without caring which specific class `feature` happens to be.